pkg/services/ngalert/state/manager.go

```diff
@@ -99,11 +110,71 @@ func (st *Manager) Run(ctx context.Context) error {
 		case <-ctx.Done():
 			st.log.Debug("Stopping")
 			ticker.Stop()
+			if st.saveStateAsync {
+				st.stateRunnerShutdown <- struct{}{}
+				st.stateRunnerWG.Wait()
+			}
 			return ctx.Err()
 		}
 	}
 }
 
+func (st *Manager) startSync(ctx context.Context) {
```

Comment:
I wasn't sure if you wanted to use break labels or if it was out of necessity, but if you did want to get rid of them you can do so using defer. I tend to find this more idiomatic.

```suggestion
func (st *Manager) startSync(ctx context.Context) {
	st.stateRunnerWG.Add(1)
	go func(ctx context.Context) {
		ticker := st.clock.Ticker(time.Second * 30)
		defer func() {
			st.log.Info("State sync shut down")
			ticker.Stop()
			st.stateRunnerWG.Done()
		}()
		for {
			select {
			case <-ticker.C:
				if err := st.fullSync(ctx); err != nil {
					st.log.Error("Failed to do a full state sync to database", "err", err)
				}
			case <-st.stateRunnerShutdown:
				st.log.Info("Stopping state sync...")
				shutdownCtx, cancel := context.WithTimeout(context.Background(), time.Second*30)
				defer cancel()
				if err := st.fullSync(shutdownCtx); err != nil {
					st.log.Error("Failed to do a full state sync to database", "err", err)
				}
				return
			}
		}
	}(ctx)
}
```

Reply:
I think the break label here is more elegant than the defer.